Add support for a custom endpoint and prefix

jeffjewiss opened the pull request:
The custom endpoint allows you to use object storage that's compatible with the AWS S3 API, such as [Minio](https://minio.io/) or Digital Ocean's new product [Spaces](https://www.digitalocean.com/products/object-storage/). Everything else works the same, but you specific a url for this property. I've been able to get it to work with Spaces on this fork.

The prefix property allows you to specify a nested folder as the destination for the app instead of the root of a bucket.

Please let me know if you'd like to see anything added or changed.

Cheers!

jeffjewiss committed:
add support for custom endpoint

jeffjewiss committed:
add support for prefix

willviles commented:
@jeffjewiss - just checking out your branch and it's working nicely. Thanks!

Just one comment... Using an `endpoint` value makes `region` redundant, yet it is asserted as a required value. Perhaps look at making it an optional value and only throw an error it if an endpoint is not given?

jeffjewiss commented:
@willviles thanks for taking a look at this.

Good point! I had missed the `requiredConfig` setup. I haven't worked much with ember deploy plugins, so I wasn't familiar with that option.

It looks like both the upload and activate steps follow the same workflow of setting up an instance of the S3 SDK. Would it make sense to create a new method, something like `this.setupS3()` that checks for the existence of either `region` or `endpoint` and throws and error if neither exist, otherwise it could return the configured S3 client?

What are your thoughts on this approach?

Cheers.

willviles commented:
@jeffjewiss sounds good to me! Whilst the [`willUpload` hook](http://ember-cli-deploy.com/docs/v1.0.x/pipeline-hooks/) is for 'confirming remote servers', I think it makes more sense to setup the S3 instance before the build in either the `configure` or `setup` hooks, so any assertions are thrown immediately.

jeffjewiss committed:
Updates configuration of plugin

* allows region to be an optional config value
* checks for one of s3Client, region or endpoint to exist
* updates tests to reflect new config options
* updates readme with new config values

jeffjewiss commented:
@willviles okay great, thanks for the feedback.

I moved the changes into a combination of the `configure` and `setup` hooks after getting caught up on the documentation for ember-cli-deploy plugins. I've made `region` optional and an error is thrown if a custom `s3Client` isn't being used and one of `region` or `endpoint` isn't present

I've also updated the tests and readme to reflect the changes.

While updating the tests I noticed that it looks like neither the npm script nor the travis config are setup to run the node tests, they're just running eslint. The travis config looks like it could also use a bump as far as Node and Ember versions. Would you be interested in a separate PR to update these?

willviles commented:
@jeffjewiss looking great!

Tests seem like they need sorting out. What do you think @he9qi? Release this change and then a separate PR for the tests?

jeffjewiss commented:
Hey @willviles, just wanted to check in on this. Are you able to merge this? Should I move forward with fixing the tests in a separate PR?

Cheers

willviles commented:
@jeffjewiss I’ve not got any permissions on this repo. I’m using your branch in a production app, so like you I’m waiting on @he9qi to review and merge!

jeffjewiss commented:
@willviles okay cool, no worries. Thanks for following up and trying out the branch! Hopefully it gets reviewed and merged soon.

he9qi commented:
looks good! let me merge it.

he9qi commented:
And a separate PR for tests ;)
